File: java/expression_evaluator/src/main/java/com/qe/test/expr/ShuntingYard.java
package com.qe.test.expr;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.List;
import java.util.Map;

public class ShuntingYard {
    private static final Map<String, Integer> PRECEDENCE = Map.of(
            "+", 2, "-", 2,
            "*", 3, "/", 3, "%", 3,
            "^", 4,
            "_", 5
    );

    private static final Map<String, Associativity> ASSOCIATIVITY = Map.of(
            "+", Associativity.LEFT, "-", Associativity.LEFT,
            "*", Associativity.LEFT, "/", Associativity.LEFT, "%", Associativity.LEFT,
            "^", Associativity.RIGHT,
            "_", Associativity.RIGHT
    );

    private enum Associativity { LEFT, RIGHT }

    public List<String> toRpn(String expression) {
        if (expression == null || expression.isBlank()) {
            throw new IllegalArgumentException("Expression must not be null or blank");
        }

        List<String> tokens = tokenize(expression);
        List<String> output = new ArrayList<>();
        Deque<String> operators = new ArrayDeque<>();

        for (int i = 0; i < tokens.size(); i++) {
            String token = tokens.get(i);

            if (isNumber(token)) {
                output.add(token);
            } else if (isFunction(token)) {
                operators.push(token);
            } else if (token.equals(",")) {
                while (!operators.isEmpty() && !operators.peek().equals("(")) {
                    output.add(operators.pop());
                }
                if (operators.isEmpty()) {
                    throw new IllegalArgumentException("Mismatched parentheses or function call");
                }
            } else if (isOperator(token)) {
                if (token.equals("-")) {
                    boolean isUnary = i == 0
                            || tokens.get(i - 1).equals("(")
                            || isOperator(tokens.get(i - 1))
                            || tokens.get(i - 1).equals(",");
                    if (isUnary) {
                        operators.push("_");
                        continue;
                    }
                }
                while (!operators.isEmpty() && isOperator(operators.peek())) {
                    String top = operators.peek();
                    if ((ASSOCIATIVITY.get(token) == Associativity.LEFT
                            && PRECEDENCE.get(token) <= PRECEDENCE.get(top))
                            || (ASSOCIATIVITY.get(token) == Associativity.RIGHT
                            && PRECEDENCE.get(token) < PRECEDENCE.get(top))) {
                        output.add(operators.pop());
                    } else {
                        break;
                    }
                }
                operators.push(token);
            } else if (token.equals("(")) {
                operators.push(token);
            } else if (token.equals(")")) {
                while (!operators.isEmpty() && !operators.peek().equals("(")) {
                    output.add(operators.pop());
                }
                if (operators.isEmpty()) {
                    throw new IllegalArgumentException("Mismatched parentheses");
                }
                operators.pop(); // discard "("
                if (!operators.isEmpty() && isFunction(operators.peek())) {
                    output.add(operators.pop());
                }
            }
        }

        while (!operators.isEmpty()) {
            String op = operators.pop();
            if (op.equals("(") || op.equals(")")) {
                throw new IllegalArgumentException("Mismatched parentheses");
            }
            output.add(op);
        }

        return output;
    }

    List<String> tokenize(String expr) {
        List<String> tokens = new ArrayList<>();
        StringBuilder current = new StringBuilder();

        for (int i = 0; i < expr.length(); i++) {
            char c = expr.charAt(i);

            if (Character.isWhitespace(c)) {
                continue;
            }

            if (Character.isDigit(c) || c == '.') {
                current.append(c);
            } else {
                if (!current.isEmpty()) {
                    tokens.add(current.toString());
                    current.setLength(0);
                }
                if (isOperatorChar(c) || c == '(' || c == ')' || c == ',') {
                    tokens.add(String.valueOf(c));
                } else if (Character.isLetter(c)) {
                    current.append(c);
                } else {
                    throw new IllegalArgumentException("Unexpected character: " + c);
                }
            }
        }

        if (!current.isEmpty()) {
            tokens.add(current.toString());
        }

        return tokens;
    }

    private boolean isNumber(String token) {
        try {
            Double.parseDouble(token);
            return true;
        } catch (NumberFormatException e) {
            return false;
        }
    }

    private boolean isOperator(String token) {
        return PRECEDENCE.containsKey(token);
    }

    private boolean isFunction(String token) {
        return Character.isLetter(token.charAt(0));
    }

    private boolean isOperatorChar(char c) {
        return "+-*/%^".indexOf(c) >= 0;
    }
}

File: python/expression_evaluator/implementation/shunting_yard.py


_operators: dict[str, tuple[int, str]] = {
    "+": (2, "L"),
    "-": (2, "L"),
    "*": (3, "L"),
    "/": (3, "L"),
    "%": (3, "L"),
    "^": (4, "R"),
    "_u": (6, "R"),
}

_functions: set[str] = {
    "sin", "cos", "sqrt", "abs", "max", "min",
}


class ShuntingYard:
    def __init__(
        self,
        custom_precedence: dict[str, tuple[int, str]] | None = None,
    ) -> None:
        self._ops = dict(_operators)
        if custom_precedence:
            self._ops.update(custom_precedence)

    def tokenize(self, expression: str) -> list[str]:
        tokens: list[str] = []
        i = 0
        while i < len(expression):
            ch = expression[i]
            if ch.isspace():
                i += 1
                continue
            if ch.isdigit() or ch == ".":
                num = ch
                i += 1
                while i < len(expression) and (
                    expression[i].isdigit() or expression[i] == "."
                ):
                    num += expression[i]
                    i += 1
                tokens.append(num)
                continue
            if ch.isalpha():
                token = ch
                i += 1
                while i < len(expression) and (
                    expression[i].isalpha() or expression[i].isdigit()
                ):
                    token += expression[i]
                    i += 1
                tokens.append(token)
                continue
            if ch in self._ops or ch in "(),":
                tokens.append(ch)
                i += 1
                continue
            raise ValueError(f"Unexpected character: {ch!r}")
        return tokens

    def to_postfix(self, expression: str) -> list[str]:
        tokens = self.tokenize(expression)
        output: list[str] = []
        op_stack: list[str] = []
        last_token: str | None = None

        for token in tokens:
            if self._is_number(token):
                output.append(token)
            elif token.isalpha() and token not in _functions:
                output.append(token)
            elif token in _functions:
                op_stack.append(token)
            elif token == ",":
                while op_stack and op_stack[-1] != "(":
                    output.append(op_stack.pop())
                if not op_stack or op_stack[-1] == "(":
                    pass
            elif token == "(":
                op_stack.append(token)
            elif token == ")":
                while op_stack and op_stack[-1] != "(":
                    output.append(op_stack.pop())
                if not op_stack:
                    raise ValueError("Mismatched parentheses")
                op_stack.pop()
                if op_stack and op_stack[-1] in _functions:
                    output.append(op_stack.pop())
            elif token in self._ops:
                if (
                    token == "-"
                    and (
                        last_token is None
                        or last_token in self._ops
                        or last_token == "("
                    )
                ):
                    op_stack.append("_u")
                else:
                    while (
                        op_stack
                        and op_stack[-1] != "("
                        and op_stack[-1] not in _functions
                        and self._precedence(op_stack[-1], token)
                    ):
                        output.append(op_stack.pop())
                    op_stack.append(token)
            else:
                raise ValueError(f"Unknown token: {token!r}")
            last_token = token

        while op_stack:
            top = op_stack.pop()
            if top == "(":
                raise ValueError("Mismatched parentheses")
            output.append(top)

        return output

    def _is_number(self, token: str) -> bool:
        try:
            float(token)
            return True
        except ValueError:
            return False

    def _precedence(self, op1: str, op2: str) -> bool:
        p1, a1 = self._ops[op1]
        p2, a2 = self._ops[op2]
        if a1 == "L":
            return p1 >= p2
        return p1 > p2
